js/lightbeam.js

```diff
@@ -1,9 +1,56 @@
+// transforms the object of nested objects 'websites' into a
+// usable format for d3
+function transformData(websites) {
+  const nodes = [];
+  const links = [];
+  const firstParties = new Set();
+  const thirdParties = new Set();
+  // make first party nodes
+  for (const firstParty in websites) {
+    firstParties.add(firstParty);
+    const node = {};
+    node.hostname = firstParty;
+    node.favicon = websites[firstParty].faviconUrl || '';
+    node.party = 'first';
+    nodes.push(node);
+    // make third party nodes and links
+    if (websites[firstParty].thirdPartyRequests) {
+      for (const thirdParty in websites[firstParty].thirdPartyRequests) {
+        addWebsite(firstParty, thirdParty);
```

Comment:
Sorry I meant `addWebsite` would share code with first party in that it would for finding favicons etc not linking website nodes together.

So this code might look more like:

```
 for (const firstParty in websites) {
    firstParties.add(createWebsite(firstParty));
    // make third party nodes and links
    if (websites[firstParty].thirdPartyRequests) {
      for (const thirdParty in websites[firstParty].thirdPartyRequests) {
        firstParties.add(createWebsite(thirdParty));
        const link = {};
        link.source = firstParty.hostname;
        link.target = thirdParty.hostname;
        links.push(link);
```

```
function createWebsite(website) {
  const websiteOutput = {
    hostname: website.hostname,
    favicon: website.faviconUrl || "".
    party: "third"
  };
  if ("firstParties" in website) {
    websiteOutput.party = "first";
  }
  return websiteOutput;
}
```